libs/d3d12/main.c

```diff
@@ -269,6 +267,76 @@ HRESULT WINAPI DLLEXPORT D3D12CreateDevice(IUnknown *adapter, D3D_FEATURE_LEVEL
         "VK_KHR_xcb_surface",
 #endif
     };
+
+    pthread_mutex_lock(&instance_singleton_lock);
+    if (instance_singleton)
+    {
+        vkd3d_instance_incref(*out_instance = instance_singleton);
+        TRACE("Returning instance singleton.\n");
+        goto out_unlock;
+    }
+
+    if (!load_vulkan())
+    {
+        ERR("Failed to load Vulkan library.\n");
+        hr = E_FAIL;
+        goto out_unlock;
+    }
+
+    memset(&instance_create_info, 0, sizeof(instance_create_info));
+    instance_create_info.pfn_vkGetInstanceProcAddr = vulkan_vkGetInstanceProcAddr;
+    instance_create_info.instance_extensions = instance_extensions;
+    instance_create_info.instance_extension_count = ARRAY_SIZE(instance_extensions);
+    instance_create_info.optional_instance_extensions = NULL;
+    instance_create_info.optional_instance_extension_count = 0;
+
+    if (FAILED(hr = vkd3d_create_instance(&instance_create_info, &instance_singleton)))
+    {
+        WARN("Failed to create vkd3d instance, hr %#x.\n", hr);
+        goto out_unlock;
+    }
+
+    vkd3d_instance_incref(*out_instance = instance_singleton);
+out_unlock:
+    pthread_mutex_unlock(&instance_singleton_lock);
+    return hr;
+}
+
+#ifdef _WIN32
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+{
+    /* Only bother here if d3d12.dll is unloaded through a mechanism such as FreeLibrary.
+     * On process termination we don't care. */
+    (void)hinstDLL;
+    if (fdwReason == DLL_PROCESS_DETACH && !lpvReserved && instance_singleton)
+    {
+        TRACE("Tearing down singleton instance.\n");
+        vkd3d_instance_decref(instance_singleton);
+    }
+    return TRUE;
+}
+#else
+__attribute__((destructor))
+static void teardown_instance(void)
+{
+    TRACE("Tearing down singleton instance.\n");
+    if (instance_singleton)
+        vkd3d_instance_decref(instance_singleton);
```

Comment:
We hold a reference to the .so here through the load_once, it should be impossible for libvulkan.so to be dead here.

Reply:
is this true even during pocess teardown though?

Reply:
Could we destroy the instance once the last D3D12 device gets destroyed instead? That should be safe and still avoid excessive instance creation.

Reply:
In DllMain the lpvReserved == nullptr means it's only called on explicit FreeLibrary, so it should be fine.

Reply:
> Could we destroy the instance once the last D3D12 device gets destroyed instead?

I suppose we could ... still leads to excessive instance creation though on boot.

Reply:
Fair enough, I guess we can do this for now but I still don't like it.

If for whatever reason instance destruction was to unload another DLL we'd have a potential deadlock here. This shouldn't happen since we're loading winevulkan directly now, but I don't *really* like this, it's all kind of scary.